File: leetcode3/최원준/15. 3Sum.py
#

'''
1. 아이디어 :
- 정렬 후, 왼쪽 포인터를 두고, 나머지 두 숫자를 찾습니다.(왼쪽+1, 오른쪽끝에서 이분탐색)
- 왼쪽 포인터가 가리키는 숫자가 양수인 경우, 리턴.

2. 시간복잡도 :
    O( n^2 )

3. 자료구조/알고리즘 :
정렬, 투포인터

'''


class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        n = len(nums)
        ans = []

        for left in range(n - 2):
            left_value = nums[left]

            if left > 0 and nums[left] == nums[left - 1]:
                continue

            if left_value > 0:
                break

            mid = left + 1
            right = n - 1

            while mid < right:
                total = left_value + nums[mid] + nums[right]

                if total == 0:
                    ans.append([left_value, nums[mid], nums[right]])

                    mid += 1
                    right -= 1

                    while mid < right and nums[mid] == nums[mid - 1]: mid += 1
                    while mid < right and nums[right] == nums[right + 1]: right -= 1

                elif total < 0:
                    mid += 1
                else:
                    right -= 1

        return ans

File: leetcode3/이진희/15_3Sum.java
/*

1. 아이디어 : 투포인터, 정렬 후 for문을 통해 숫자 고정, 이후 투포인터로 값찾기

2. 시간복잡도 : O(NlogN)*O(N) -> O(N^2)

3. 자료구조/알고리즘 : 투포인터

 */

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);

        for(int i=0; i<nums.length-2; i++) {
            if(i>0 && nums[i] == nums[i-1]) continue;

            int left = i+1;
            int right = nums.length-1;

            while(left<right) {
                int sum = nums[i] + nums[left] + nums[right];
                if(sum == 0) {
                    ans.add(Arrays.asList(nums[i], nums[left], nums[right]));

                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;
                    
                    left++;
                    right--;
                }
                else if(sum < 0) left++;
                else right--;
            }
        }
        return ans;
    }
}